modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java

```diff
@@ -190,13 +190,23 @@ private void failQueuedWrites() {
 
     private static final class WriteOperation {
 
-        private final ByteBuf buf;
+        private ByteBuf buf;
+
+        private OutboundHandler.SendContext context;
 
         private final ChannelPromise promise;
 
-        WriteOperation(ByteBuf buf, ChannelPromise promise) {
-            this.buf = buf;
+        WriteOperation(OutboundHandler.SendContext context, ChannelPromise promise) {
+            this.context = context;
             this.promise = promise;
         }
+
+        ByteBuf buffer() throws IOException {
+            if (buf == null) {
+                buf = Netty4Utils.toByteBuf(context.get());
+                context = null;
+            }
```

Comment:
Can we `assert context == null` here?

Reply:
Sure thing, added it. But yes, all of this is happening on a single IO thread thread so this is all safe I think :)